src/cpp/shared/town/town.cpp

```diff
@@ -12,19 +12,175 @@
 #include "spell/spells.h"
 #include "town/town.h"
 
+#include <cstdio>
 #include <sstream>
 #include <string>
+#include <vector>
 
-unsigned long gTownEligibleBuildMask[] = {
+unsigned long gTownEligibleBuildMask[NUM_FACTIONS] = {
   0x3FF8BF9F,
   0x1BF8BF9F,
   0xFF8BF9F,
   0x69F8BF9F,
   0x35F8BF9F,
-  0x1FF8BF9B
+  0x1FF8BF9B,
+  0x0,
+  0x0,
+  0x0,
+  0x0,
+  0x0,
+  0x0,
+  0x01F8BF9F  // TODO: decide whether Cyborg creatures have any upgrades
 };
 
+// Leave these undefined initially to make the compiler put them in the
+// same data segment. This ensures the old game code will be able to index
```

Comment:
All such code was in SetupTowns, which you've fixed, correct? So you don't need to leave these undefined initially, right?

Reply:
Actually, I do.  There's something about initializing `gTownObjNames` with pointers that it ends up in a different memory segment from `gDwellingType`.  Leaving them both uninitialized was the only way I found to make them appear consecutively.

Reply:
But the only reason to make them appear consecutively is the OOB accesses of `gTownObjNames`, right? And we can eliminate all such OOB accesses, right?

And yes, that makes it change segments. Unitialized arrays are in the .BSS section (which does not stand for "Better Save Space," but might as well). Initialized arrays are in the .DATA section. Const initialized arrays are in the .RDATA section; however, the version of MS C++ used to compile the original game put these in the .DATA section.